src/Microsoft.OData.Core/Evaluation/ODataConventionalEntityMetadataBuilder.cs

```diff
@@ -408,7 +408,11 @@ private Uri ComputeEditLink()
             Debug.Assert(this.ResourceMetadataContext != null && this.ResourceMetadataContext.TypeContext != null, "this.resourceMetadataContext != null && this.resourceMetadataContext.TypeContext != null");
             if (this.ResourceMetadataContext.ActualResourceTypeName != this.ResourceMetadataContext.TypeContext.NavigationSourceEntityTypeName)
             {
-                uri = this.UriBuilder.AppendTypeSegment(uri, this.ResourceMetadataContext.ActualResourceTypeName);
+                // In some scenarios, the resource Id will already contain the cast segment
+                if (!uri.ToString().EndsWith(string.Concat('/', this.ResourceMetadataContext.ActualResourceTypeName), StringComparison.Ordinal))
+                {
+                    uri = this.UriBuilder.AppendTypeSegment(uri, this.ResourceMetadataContext.ActualResourceTypeName);
+                }

```

Comment:
I considered using the following logic as an alternative but was concerned about increase in allocations:
```csharp
var odataUriParser = new OData.UriParser.ODataUriParser(this.MetadataContext.Model, this.MetadataContext.ServiceBaseUri, uri);
var odataPath = odataUriParser.ParsePath();
if (!(odataPath.LastSegment is OData.UriParser.TypeSegment typeSegment && typeSegment.TargetEdmType == this.ResourceMetadataContext.ActualResourceType))
{
	uri = this.UriBuilder.AppendTypeSegment(uri, this.ResourceMetadataContext.ActualResourceTypeName);
}
```

Reply:
Eventually adopted this approach since it made it easy to cater for the scenario raised [here](https://github.com/OData/odata.net/pull/2592/files#r1080945648)